File: TextSegmentation/js/js/scenario2-getcurrenttextsegmentfromindex.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";
    var page = WinJS.UI.Pages.define("/html/scenario2-GetCurrentTextSegmentFromIndex.html", {
        ready: function (element, options) {
            document.getElementById("WordSegmentButton").addEventListener("click", showWordSegmentFromIndex, false);
            document.getElementById("SelectionSegmentButton").addEventListener("click", showSelectionSegmentFromIndex, false);
        }
    });

    function showWordSegmentFromIndex() {

        // Initialize and obtain input values
        var notifyText = "";

        // Obtain the input string value, check for non-emptiness
        var inputStringBox = document.getElementById("inputStringBox");
        var inputStringText = inputStringBox.value;
        if ("" === inputStringText) {
            notifyText = "\nCannot compute word segments: input string is empty.";
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }

        // Obtain the language tag value, check for non-emptiness and that it is valid
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        var languageTagBox = document.getElementById("languageTagBox");
        var languageTagText = languageTagBox.value;
        if ("" === languageTagText) {
            notifyText += "\nLanguage tag input is empty ... using generic-language segmentation rules.";
            languageTagText = "und";
        }
        else {
            if (!Windows.Globalization.Language.isWellFormed(languageTagText)) {
                notifyText = "\nLanguage tag is not well formed: \"" + languageTagText + "\"";
                WinJS.log && WinJS.log(notifyText, "sample", "error");
                return;
            }
        }

        // Obtain the text index
        //   : If empty, assume the default ( = 0), returns the first segment
        //   : If not an actual number, throw an error
        //   : If Negative, or beyond the length of the input string, throw an error
        var textIndex = textIndexBox.value;
        if ("" === textIndex) {
            textIndex = 0;
            notifyText += "\nNo input index provided ... using first segment reference (index = 0) as default.";
        }
        else if (isNaN(parseFloat(textIndex)) || !isFinite(textIndex)) {
            notifyText = "\nIndex value must be a valid number";
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }
        else if (textIndex < 0) {
            notifyText = "\nCannot compute word segment for negative index value: " + textIndex;
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }
        else if (textIndex >= inputStringText.length) {
            notifyText = "\nCannot compute word segment for out of bounds index value: " + textIndex;
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }

        // Notify that we are going to calculate selection segment
        notifyText += "\n\nFinding the word segment for the given index ...\n";
        notifyText += "\nInput: \"" + inputStringText + "\"";
        notifyText += "\nLanguage Tag: \"" + languageTagText + "\"";
        notifyText += "\nIndex: " + textIndex + "\n";

        // Construct the WordsSegmenter instance
        var segmenter = new Windows.Data.Text.WordsSegmenter(languageTagText);

        // Obtain the token segment
        var tokenSegment = segmenter.getTokenAt(inputStringText, textIndex);
        notifyText += "\nIndexed segment: \"" + tokenSegment.text + "\"";

        // send notifyText to the output pane
        WinJS.log && WinJS.log(notifyText, "sample", "status");
    }

    function showSelectionSegmentFromIndex() {

        // Initialize and obtain input values
        var notifyText = "";

        // Obtain the input string value, check for non-emptiness
        var inputStringBox = document.getElementById("inputStringBox");
        var inputStringText = inputStringBox.value;
        if ("" === inputStringText) {
            notifyText = "\nCannot compute selection segments: input string is empty.";
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }

        // Obtain the language tag value, check for non-emptiness and that it is valid
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        var languageTagBox = document.getElementById("languageTagBox");
        var languageTagText = languageTagBox.value;
        if ("" === languageTagText) {
            notifyText += "\nLanguage tag input is empty ... using generic-language segmentation rules.";
            languageTagText = "und";
        }
        else {
            if (!Windows.Globalization.Language.isWellFormed(languageTagText)) {
                notifyText = "\nLanguage tag is not well formed: \"" + languageTagText + "\"";
                WinJS.log && WinJS.log(notifyText, "sample", "error");
                return;
            }
        }

        // Obtain the text index
        //   : If empty, assume the default ( = 0), returns the first segment
        //   : If not an actual number, throw an error
        //   : If Negative, or beyond the length of the input string, throw an error
        var textIndex = textIndexBox.value;
        if ("" === textIndex) {
            textIndex = 0;
            notifyText += "\nNo input index provided ... using first segment reference (index = 0) as default.";
        }
        else if (isNaN(parseFloat(textIndex)) || !isFinite(textIndex)) {
            notifyText = "\nIndex value must be a valid number";
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }
        else if (textIndex < 0) {
            notifyText = "\nCannot compute selection segment for negative index value: " + textIndex;
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }
        else if (textIndex >= inputStringText.length) {
            notifyText = "\nCannot compute selection segment for out of bounds index value: " + textIndex;
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }

        // Notify that we are going to calculate selection segment
        notifyText += "\n\nFinding the selection segment for the given index ...\n";
        notifyText += "\nInput: \"" + inputStringText + "\"";
        notifyText += "\nLanguage Tag: \"" + languageTagText + "\"";
        notifyText += "\nIndex: " + textIndex + "\n";

        // Construct the SelectableWordsSegmenter instance
        var segmenter = new Windows.Data.Text.SelectableWordsSegmenter(languageTagText);

        // Obtain the token segment
        var tokenSegment = segmenter.getTokenAt(inputStringText, textIndex);
        notifyText += "\nToken segment: \"" + tokenSegment.text + "\"";

        // send notifyText to the output pane
        WinJS.log && WinJS.log(notifyText, "sample", "status");
    }
})();

File: Samples/TextSegmentation/cppwinrt/Scenario2_GetCurrentTextSegmentFromIndex.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario2_GetCurrentTextSegmentFromIndex.h"
#include "Scenario2_GetCurrentTextSegmentFromIndex.g.cpp"

using namespace winrt;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;

namespace winrt::SDKTemplate::implementation
{
    Scenario2_GetCurrentTextSegmentFromIndex::Scenario2_GetCurrentTextSegmentFromIndex()
    {
        InitializeComponent();
    }

    // This is the click handler for the 'Word Segment' button.
    //
    // When this button is activated, the Text Segmentation API will calculate
    // the word segment from the given input string and character index for that string, 
    // and return the WordSegment object that contains the index within its text bounds.
    // Segment breaking behavior is based off of the language-tag input, which defines 
    // which language rules to use.
    void Scenario2_GetCurrentTextSegmentFromIndex::WordSegmentButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        std::wostringstream result;

        // Obtain the input string value, check for non-emptiness
        hstring inputStringText = inputStringBox().Text();
        if (inputStringText.empty())
        {
            rootPage.NotifyUser(L"Cannot compute word segments: input string is empty.", NotifyType::ErrorMessage);
            return;
        }

        // Obtain the language tag value, check for non-emptiness
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        hstring languageTagText = languageTagBox().Text();
        if (languageTagText.empty())
        {
            result << L"Language tag input is empty ... using generic-language segmentation rules.\n";
            languageTagText = L"und";   // This is used for non language-specific locales. 'und' is short for 'undetermined'.
        }
        else
        {
            if (!Windows::Globalization::Language::IsWellFormed(languageTagText))
            {
                rootPage.NotifyUser(L"Language tag is not well formed: \"" + languageTagText + L"\"", NotifyType::ErrorMessage);
                return;
            }
        }

        // Obtain the input Index
        hstring inputIndexString = indexBox().Text();
        uint32_t inputIndex = static_cast<uint32_t>(_wtoi(inputIndexString.c_str()));
        if (errno == EINVAL)
        {
            result << L"No input index provided ... using first segment reference (index = 0) as default.\n";
        }
        else if (inputIndex > inputStringText.size())
        {
            rootPage.NotifyUser(L"Invalid index supplied ... cannot use a negative index, or an index that is out of bounds of the input string.\n\nPlease re-check the index value, and try again.",
                NotifyType::ErrorMessage);
            return;
        }

        // Notify that we are going to calculate word segment
        result
            << L"\n\nFinding the word segment for the given index ...\n"
            << L"Input: \"" << std::wstring_view(inputStringText) << L"\"\n"
            << L"Language Tag: \"" << std::wstring_view(languageTagText) << L"\"\n"
            << L"Index: " << inputIndex << L"\n";

        // Construct the WordsSegmenter instance
        Windows::Data::Text::WordsSegmenter segmenter(languageTagText);

        // Obtain the token segment
        auto tokenSegment = segmenter.GetTokenAt(inputStringText, inputIndex);
        result << L"\nIndexed segment: \"" << std::wstring_view(tokenSegment.Text()) << L"\"";

        // Set results to output box
        rootPage.NotifyUser(result.str(), NotifyType::StatusMessage);
    }

    // This is the click handler for the 'Selection Segment' button.
    //
    // When this button is activated, the Text Segmentation API will calculate
    // the selection segment from the given input string and character index for that string, 
    // and return the SelectableWordSegment object that contains the index within its text bounds.
    // Segment breaking behavior is based off of the language-tag input, which defines 
    // which language rules to use.
    //
    // Selection segments differ from word segments in that they describe the bounds
    // between active-selection snapping behavior.
    void Scenario2_GetCurrentTextSegmentFromIndex::SelectionSegmentButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        std::wostringstream result;

        // Obtain the input string value, check for non-emptiness
        hstring inputStringText = inputStringBox().Text();
        if (inputStringText.empty())
        {
            rootPage.NotifyUser(L"Cannot compute word segments: input string is empty.", NotifyType::ErrorMessage);
            return;
        }

        // Obtain the language tag value, check for non-emptiness
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        hstring languageTagText = languageTagBox().Text();
        if (languageTagText.empty())
        {
            result << L"Language tag input is empty ... using generic-language segmentation rules.\n";
            languageTagText = L"und";   // This is used for non language-specific locales. 'und' is short for 'undetermined'.
        }
        else
        {
            if (!Windows::Globalization::Language::IsWellFormed(languageTagText))
            {
                rootPage.NotifyUser(L"Language tag is not well formed: \"" + languageTagText + L"\"", NotifyType::ErrorMessage);
                return;
            }
        }

        // Obtain the input Index
        hstring inputIndexString = indexBox().Text();
        uint32_t inputIndex = static_cast<uint32_t>(_wtoi(inputIndexString.c_str()));
        if (errno == EINVAL)
        {
            result << L"No input index provided ... using first segment reference (index = 0) as default.\n";
        }
        else if (inputIndex > inputStringText.size())
        {
            rootPage.NotifyUser(L"Invalid index supplied ... cannot use a negative index, or an index that is out of bounds of the input string.\n\nPlease re-check the index value, and try again.",
                NotifyType::ErrorMessage);
            return;
        }

        // Notify that we are going to calculate word segment
        result
            << L"\n\nFinding the selection segment for the given index ...\n"
            << L"Input: \"" << std::wstring_view(inputStringText) << L"\"\n"
            << L"Language Tag: \"" << std::wstring_view(languageTagText) << L"\"\n"
            << L"Index: " << inputIndex << L"\n";

        // Construct the SelectableWordsSegmenter instance
        Windows::Data::Text::SelectableWordsSegmenter segmenter(languageTagText);

        // Obtain the token segment
        auto tokenSegment = segmenter.GetTokenAt(inputStringText, inputIndex);
        result << L"\nIndexed segment: \"" << std::wstring_view(tokenSegment.Text()) << L"\"";

        // Set results to output box
        rootPage.NotifyUser(result.str(), NotifyType::StatusMessage);
    }
}
